substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java

```diff
@@ -156,23 +169,35 @@ private void addLineNumberRecords(CompiledMethodEntry compiledEntry) {
      *
      * @param record the symbol subrecord to add.
      */
-    private void addToSymbolSubsection(CVSymbolSubrecord record) {
+    private void addSymbolRecord(CVSymbolSubrecord record) {
         cvSymbolSubsection.addRecord(record);
     }
 
     /**
-     * Add type records for function. (later add arglist, and return type and local types).
+     * Add type records for a class and all its members.
      *
-     * @param entry compild method containing entities whoses type records must be added
-     * @return type index of function type
+     * @param typeEntry class to add records for.
      */
-    private int addTypeRecords(@SuppressWarnings("unused") CompiledMethodEntry entry) {
-        CVTypeRecord.CVTypeArglistRecord argListType = addTypeRecord(new CVTypeRecord.CVTypeArglistRecord().add(T_NOTYPE));
-        CVTypeRecord funcType = addTypeRecord(new CVTypeRecord.CVTypeProcedureRecord().returnType(T_VOID).argList(argListType));
-        return funcType.getSequenceNumber();
+    private void addTypeRecords(TypeEntry typeEntry) {
+        int typeIdx = cvDebugInfo.getCVTypeSection().addTypeRecords(typeEntry).getSequenceNumber();
+
+        if (typeIdx > MAX_PRIMITIVE) {
+            /*
+             * Adding an S_UDT (User Defined Type) record ensures the linker doesn't throw away the
+             * type definition.
+             */
+            CVSymbolSubrecord.CVSymbolUDTRecord udtRecord = new CVSymbolSubrecord.CVSymbolUDTRecord(cvDebugInfo, typeIdx, CVNames.typeNameToCodeViewName(typeEntry.getTypeName()));
+            addSymbolRecord(udtRecord);
+        }
     }
 
-    private <T extends CVTypeRecord> T addTypeRecord(T record) {
-        return cvDebugInfo.getCVTypeSection().addOrReference(record);
+    /**
+     * Add type records for a class and all its members.
+     *
+     * @param entry compiled method containing entities whos type records must be added
+     * @return type index of function type
+     */
+    private int addTypeRecords(@SuppressWarnings("unused") CompiledMethodEntry entry) {
```

Comment:
This is no longer necessary.
```suggestion
    private int addTypeRecords(CompiledMethodEntry entry) {
```

Reply:
Done.